load data and store in dataframe df:

In [1]:
import pandas as pd
import random
from sklearn.linear_model import LinearRegression

In [47]:
# path of data 
path = 'Automobile_data.csv'
df = pd.read_csv(path, na_values = ["?"], keep_default_na = False) # all ? values will be Nulls
df.head() 

symboling  normalized-losses         make fuel-type aspiration  \
0          3                NaN  alfa-romero       gas        std   
1          3                NaN  alfa-romero       gas        std   
2          1                NaN  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0          130         mpfi  3.47    2.68               9.0      111.0   
1          130         mpfi  3.47    2.68               9.0      111.0   
2          152         mpfi  2.68    3.47               9.0      154.0   
3          109         mpfi  3.19    3.40              10.0      102.0   
4          136         mpfi  3.19    3.40               8.0      115.0   

   peak-rpm city-mpg  highway-mpg    price  
0    5000.0       21           27  13495.0  
1    5000.0       21           27  16500.0  
2    5000.0       19           26  16500.0  
3    5500.0       24           30  13950.0  
4    5500.0       18           22  17450.0  

[5 rows x 26 columns]

In [48]:
df= df[pd.to_numeric(df["price"], errors='coerce').notnull()] # delete rows with Nulls
df= df[pd.to_numeric(df["horsepower"], errors='coerce').notnull()]

def gauss_with_mean(x):
    try:
        x = int(x)
        return x
    except ValueError:
        return  m + random.gauss(0,1)

# because normalized lossed has to many Nulls we will replace them with the mean value + some random deviation
    
m = df["normalized-losses"].mean()

df["normalized-losses"] = df["normalized-losses"].apply(gauss_with_mean)

In [49]:
df.reset_index(inplace = True)

In [50]:
df.head(n = 10) 

index  symboling  normalized-losses         make fuel-type aspiration  \
0      0          3         121.123619  alfa-romero       gas        std   
1      1          3         123.327872  alfa-romero       gas        std   
2      2          1         122.876431  alfa-romero       gas        std   
3      3          2         164.000000         audi       gas        std   
4      4          2         164.000000         audi       gas        std   
5      5          2         121.385571         audi       gas        std   
6      6          1         158.000000         audi       gas        std   
7      7          1         123.927483         audi       gas        std   
8      8          1         158.000000         audi       gas      turbo   
9     10          2         192.000000          bmw       gas        std   

  num-of-doors   body-style drive-wheels engine-location  ...  engine-size  \
0          two  convertible          rwd           front  ...          130   
1          two  convertible          rwd           front  ...          130   
2          two    hatchback          rwd           front  ...          152   
3         four        sedan          fwd           front  ...          109   
4         four        sedan          4wd           front  ...          136   
5          two        sedan          fwd           front  ...          136   
6         four        sedan          fwd           front  ...          136   
7         four        wagon          fwd           front  ...          136   
8         four        sedan          fwd           front  ...          131   
9          two        sedan          rwd           front  ...          108   

   fuel-system  bore  stroke  compression-ratio horsepower peak-rpm  city-mpg  \
0         mpfi  3.47    2.68                9.0      111.0   5000.0        21   
1         mpfi  3.47    2.68                9.0      111.0   5000.0        21   
2         mpfi  2.68    3.47                9.0      154.0   5000.0        19   
3         mpfi  3.19    3.40               10.0      102.0   5500.0        24   
4         mpfi  3.19    3.40                8.0      115.0   5500.0        18   
5         mpfi  3.19    3.40                8.5      110.0   5500.0        19   
6         mpfi  3.19    3.40                8.5      110.0   5500.0        19   
7         mpfi  3.19    3.40                8.5      110.0   5500.0        19   
8         mpfi  3.13    3.40                8.3      140.0   5500.0        17   
9         mpfi  3.50    2.80                8.8      101.0   5800.0        23   

  highway-mpg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  
5          25  15250.0  
6          25  17710.0  
7          25  18920.0  
8          20  23875.0  
9          29  16430.0  

[10 rows x 27 columns]

In [51]:
from sklearn.linear_model import LinearRegression

In [52]:
lm = LinearRegression()

In [53]:
size = int(0.2 * df.shape[0])

In [54]:
test_database = pd.DataFrame() # lets create a test set and a training set manually

In [55]:
test_database

Empty DataFrame
Columns: []
Index: []

In [56]:
seen = set()
to_drop = []
while size:
    ind = random.randint(min(df.index), max(df.index))
    #print(ind)
    while ind in seen:
        ind = random.randint(min(df.index), max(df.index))
    seen.add(ind)
    to_drop.append(ind)
    row = df.iloc[ind]
    test_database = pd.concat([test_database, pd.DataFrame([row])])
    size -= 1
else:
    df.drop(to_drop, inplace = True)

In [57]:
test_database.reset_index(inplace = True)
df.reset_index(inplace = True)

In [58]:
df.head(n = 10)

level_0  index  symboling  normalized-losses         make fuel-type  \
0        0      0          3         121.123619  alfa-romero       gas   
1        1      1          3         123.327872  alfa-romero       gas   
2        2      2          1         122.876431  alfa-romero       gas   
3        3      3          2         164.000000         audi       gas   
4        4      4          2         164.000000         audi       gas   
5        5      5          2         121.385571         audi       gas   
6        6      6          1         158.000000         audi       gas   
7        7      7          1         123.927483         audi       gas   
8        8      8          1         158.000000         audi       gas   
9        9     10          2         192.000000          bmw       gas   

  aspiration num-of-doors   body-style drive-wheels  ... engine-size  \
0        std          two  convertible          rwd  ...         130   
1        std          two  convertible          rwd  ...         130   
2        std          two    hatchback          rwd  ...         152   
3        std         four        sedan          fwd  ...         109   
4        std         four        sedan          4wd  ...         136   
5        std          two        sedan          fwd  ...         136   
6        std         four        sedan          fwd  ...         136   
7        std         four        wagon          fwd  ...         136   
8      turbo         four        sedan          fwd  ...         131   
9        std          two        sedan          rwd  ...         108   

   fuel-system  bore  stroke  compression-ratio  horsepower peak-rpm city-mpg  \
0         mpfi  3.47    2.68                9.0       111.0   5000.0       21   
1         mpfi  3.47    2.68                9.0       111.0   5000.0       21   
2         mpfi  2.68    3.47                9.0       154.0   5000.0       19   
3         mpfi  3.19    3.40               10.0       102.0   5500.0       24   
4         mpfi  3.19    3.40                8.0       115.0   5500.0       18   
5         mpfi  3.19    3.40                8.5       110.0   5500.0       19   
6         mpfi  3.19    3.40                8.5       110.0   5500.0       19   
7         mpfi  3.19    3.40                8.5       110.0   5500.0       19   
8         mpfi  3.13    3.40                8.3       140.0   5500.0       17   
9         mpfi  3.50    2.80                8.8       101.0   5800.0       23   

   highway-mpg    price  
0           27  13495.0  
1           27  16500.0  
2           26  16500.0  
3           30  13950.0  
4           22  17450.0  
5           25  15250.0  
6           25  17710.0  
7           25  18920.0  
8           20  23875.0  
9           29  16430.0  

[10 rows x 28 columns]

In [59]:
test_database.head(n = 10)

level_0  index  symboling  normalized-losses           make fuel-type  \
0      145    151          1          87.000000         toyota       gas   
1      172    178          3         197.000000         toyota       gas   
2      178    184          2          94.000000     volkswagen    diesel   
3       36     37          0         106.000000          honda       gas   
4      112    115          0         161.000000         peugot       gas   
5       27     28         -1         110.000000          dodge       gas   
6       63     66          0         121.766046          mazda    diesel   
7       31     32          1         101.000000          honda       gas   
8       64     67         -1          93.000000  mercedes-benz    diesel   
9       60     63          0         121.216672          mazda    diesel   

  aspiration num-of-doors body-style drive-wheels  ... engine-size  \
0        std          two  hatchback          fwd  ...          92   
1        std          two  hatchback          rwd  ...         171   
2        std         four      sedan          fwd  ...          97   
3        std          two  hatchback          fwd  ...         110   
4        std         four      sedan          rwd  ...         120   
5        std         four      wagon          fwd  ...         122   
6        std         four      sedan          rwd  ...         134   
7        std          two  hatchback          fwd  ...          79   
8      turbo         four      sedan          rwd  ...         183   
9        std          NaN      sedan          fwd  ...         122   

   fuel-system  bore  stroke  compression-ratio  horsepower peak-rpm city-mpg  \
0         2bbl  3.05    3.03                9.0        62.0   4800.0       31   
1         mpfi  3.27    3.35                9.3       161.0   5200.0       20   
2          idi  3.01    3.40               23.0        52.0   4800.0       37   
3         1bbl  3.15    3.58                9.0        86.0   5800.0       27   
4         mpfi  3.46    3.19                8.4        97.0   5000.0       19   
5         2bbl  3.34    3.46                8.5        88.0   5000.0       24   
6          idi  3.43    3.64               22.0        72.0   4200.0       31   
7         1bbl  2.91    3.07               10.1        60.0   5500.0       38   
8          idi  3.58    3.64               21.5       123.0   4350.0       22   
9          idi  3.39    3.39               22.7        64.0   4650.0       36   

   highway-mpg    price  
0           38   6338.0  
1           24  16558.0  
2           46   7995.0  
3           33   7895.0  
4           24  16630.0  
5           30   8921.0  
6           39  18344.0  
7           42   5399.0  
8           25  25552.0  
9           42  10795.0  

[10 rows x 28 columns]

In [60]:
X = df[["horsepower", "curb-weight", "engine-size", "highway-mpg"]]
Y = df[['price']]

In [61]:
lm.fit(X, Y)

LinearRegression()

In [62]:
X = test_database[["horsepower", "curb-weight", "engine-size", "highway-mpg"]]
Y = test_database[['price']]

In [63]:
Yha =lm.predict(X)
Yha[0:5]   

array([[ 5710.30741416],
       [21936.21109337],
       [ 6197.99183863],
       [ 9480.24718817],
       [14691.48123975]])

In [64]:
Y.head() # predictions are not perfect, but relatively close

price
0   6338.0
1  16558.0
2   7995.0
3   7895.0
4  16630.0

In [65]:
lm.score(X,Y) # R**2

0.7940336324890291

In [66]:
lm.intercept_ # intercept

array([-12231.33017938])

In [67]:
lm.coef_ # coefficents

array([[ 55.88920659,   4.14697557,  79.96806301, -35.27329403]])

Equation:

Y = 55.88920659 * x1 + 4.14697557 * x2 + 79.96806301 * x3 -35.27329403 * x4 -12231.33017938

In [68]:
X = df[["normalized-losses","highway-mpg"]]
Y = df[['price']]

In [69]:
lm2=LinearRegression() # second model
lm2 

LinearRegression()

In [70]:
lm2.fit(X,Y) # train the model

LinearRegression()

In [71]:
X = test_database[["normalized-losses","highway-mpg"]]
Y = test_database[['price']]

In [72]:
Yha =lm2.predict(X)
Yha[0:5] 

array([[ 6690.67757276],
       [18575.00055078],
       [ -207.0027177 ],
       [10966.04526129],
       [18629.95540632]])

In [73]:
Y.head()

price
0   6338.0
1  16558.0
2   7995.0
3   7895.0
4  16630.0

In [74]:
lm2.coef_ #coefficents

array([[  -1.52652377, -860.87432801]])

In [75]:
lm2.intercept_ # intercept

array([39536.70960484])

In [76]:
lm2.score(X,Y) # R**2

0.44787248702635507

Equation

Y = -1.52652377*x1 - 860.87432801 *x2 + 34314.44215086

Note that if you run the file again, all models will be different! This is because we randomly chose test and train sets, and we do not have a lot of data here, so these changes affect the models signifficantly. In such case it is probably better to either add more data or use k-folds to have a better estimation of our equations